Support running Studio with optimized assets

andy-armstrong opened the pull request:
This is a follow-up to https://github.com/edx/edx-platform/pull/8529 which updated Bok Choy to run using optimized assets. This change allows developers to run Studio in devstack with optimized assets, using a new devstack_optimized settings file.

The following Paver command can be used to run Studio in optimized mode:

    paver devstack studio --optimized

You can also generate the assets explicitly and then run Studio:

    paver update_assets cms --settings=test_static_optimized
    paver devstack studio --settings=devstack_optimized --fast

@cahrens @jzoldak please review.

andy-armstrong commented:
Jenkins, test this please

cahrens commented:
Awesome! I'll take a look at this tomorrow.

jzoldak commented on cms/envs/devstack_optimized.py:
```diff
@@ -0,0 +1,51 @@
+  paver devstack studio --settings=devstack_optimized --fast
+
+Note that changes to JavaScript assets will not be picked up automatically
+as they are for devstack. Instead, update_assets must be invoked each time
```
as they are for non-optimized devstack.

cahrens commented on cms/envs/devstack_optimized.py:
```diff
@@ -0,0 +1,51 @@
+that changes have been made.
+"""
+
+# We intentionally define lots of variables that aren't used, and
```
I think you can remove this comment block.

cahrens commented on cms/envs/devstack_optimized.py:
```diff
@@ -0,0 +1,51 @@
+# want to import all variables from base settings files
+# pylint: disable=wildcard-import, unused-wildcard-import
+
+# Pylint gets confused by path.py instances, which report themselves as class
```
I would get rid of this comment block too and just put the pylint disable statement on line 30 (pylint: disable=no-name-in-module).

cahrens commented on cms/envs/devstack_optimized.py:
```diff
@@ -0,0 +1,51 @@
+
+########################## Devstack settings ###################################
+
+from .devstack import *  # pylint: disable=wildcard-import, unused-wildcard-import
```
I have a feeling we are going to have an exact copy of this file in lms/envs, which makes me sad.

andy-armstrong replied:
Yes, we are. :-( Let me know if you can think of a good way to share these settings.

cahrens commented on pavelib/servers.py:
```diff
@@ -28,9 +28,10 @@ def run_server(system, settings=None, port=None, skip_assets=False, contracts=Fa
-        # Local dev settings use staticfiles to serve assets, so we can skip the collecstatic step
-        args = [system, '--settings={}'.format(settings), '--skip-collect', '--watch']
+    if asset_settings:
+        args = [system, '--settings={}'.format(asset_settings), '--skip-collect', '--watch']
```
I'm confused about "--skip-collect", which appears on lines 32 and 34. You might need to explain this to me (if this code is actually correct).

Is it necessary to have both collect_static and asset_settings?

andy-armstrong replied:
My understanding is that ``--skip-collect`` only skips the call to collectstatic but does all of the other asset processing (SASS and Coffee compilation etc). Before my changes, this code assumed that it could always skip collectstatic because the servers didn't run with those assets. However, with this change, we do want to collectstatic if we are running devstack with optimized assets.

Given this, the distinction is that you can disable all asset updating by passing None for ``asset_settings``, while you can skip just the static asset collection by passing ``collect_static=False``. I had attempted to describe this in the docstrings, so let me know if you think it needs more detail.

cahrens commented on pavelib/servers.py:
```diff
@@ -89,6 +90,8 @@ def devstack(args):
     parser.add_argument('system', type=str, nargs=1, help="lms or studio")
     parser.add_argument('--fast', action='store_true', default=False, help="Skip updating assets")
+    parser.add_argument('--optimized', action='store_true', default=False, help="Run with optimized assets")
+    parser.add_argument('--settings', type=str, default="devstack", help="Settings file")
```
Something for asset_settings?

andy-armstrong replied:
I was thinking of ``asset_settings`` as an internal option that I used to implement the optimized mode. However, it makes sense to support it externally too, so I've added it to all the server methods.

jzoldak commented:
Some questions/comments that I have:
* I'm not sure why the devstack task is marked no-help. Is it because we expect that people use the `paver lms` and `paver cms` commands instead? We should check the documentation and possibly survey other developers to see what they use. I'd hesitate to implement this under a command that when you try to run with --help starts up the server rather than give you help.
* At any rate this new optimization should be documented somewhere, perhaps on the [Developing on the edX Developer Stack wiki page](../wiki/Developing-on-the-edX-Developer-Stack)?
* Will this work with theming? If not, a note to that effect might be helpful.
* Tests should be added to our library of [paver tests](../tree/master/pavelib/paver_tests) to verify that the options are all wired through correctly.

cahrens commented:
Running the individual commands (optimize first, then run) works, but running the combined command does not actually do the optimization.

andy-armstrong commented:
@cahrens @jzoldak Thanks for the great feedback. I've addressed all of your comments, and I've also introduced tests for all the Paver server commands (both the ones I've changed and the ones that I've left untouched). Please re-review.

@jzoldak I plan on updating the wiki once this has landed. I also intend to write up a page describing all the details of the pipeline that @cahrens and I have learned these past couple of weeks.

cahrens commented on pavelib/servers.py:
```diff
@@ -2,24 +2,28 @@
     Start the server for the specified `system` (lms or studio).
     `settings` is the Django settings module to use; if not provided, use the default.
+    `collect_static` defaults to False, but if True then static files will be collected.
+    `asset_settings` is the settings to use when generating assets; if not provided, assets are not generated.
```
Does it make any sense to supply asset_settings but set collect_static to False? I'm still trying to understand how these 2 options interact.

andy-armstrong replied:
The asset_settings are used for more than just collectstatic. They are used for generating SASS and coffee, amongst other asset handling. Collecting the static files for production is only one part of what update_assets does.

The purpose of updating assets without collecting static is when you know that the server you are starting isn't going to use the collected static assets. In general, if you are using devstack then there's no need to collect static, but if you use a different settings file then you might want to. Currently the only settings file that needs the collected assets is my new devstack_optimized, but I didn't want to assume that that was the case.

I should update the documentation to make this all clearer.

cahrens commented on pavelib/servers.py:
```diff
@@ -2,24 +2,28 @@
     `port` is the port to run the server on; if not provided, use the default port for the system.
-
-    If `skip_assets` is True, skip the asset compilation step.
+    `no_contracts` is true if contracts are not to be enabled. The default is to include contracts.
```
What are contracts?

andy-armstrong replied:
I believe that this is related to enabling PyContracts in the code:

https://andreacensi.github.io/contracts/

This already existed, but I just standardized the behavior (my tests discovered that they were inconsistent about what the default was and whether the option was even supported).

cahrens commented on pavelib/servers.py:
```diff
@@ -27,10 +31,12 @@ def run_server(system, settings=None, port=None, skip_assets=False, contracts=Fa
 
     if not settings:
         settings = DEFAULT_SETTINGS
+        asset_settings = settings
```
Is this line correct? Why are we setting asset_setting to a DEFAULT_SETTINGS, and only in the case when no settings are provided? Wouldn't it make more sense to leave asset_settings undefined?

andy-armstrong replied:
Good point. The problem was that some callers assumed that they didn't have to specify any settings and the defaults would work, but then the assets wouldn't get updated. I'll try removing this line and see what else I have to fix up to make it consistent.

andy-armstrong replied:
I removed that line and all the tests carried on passing, so I think it isn't necessary.

cahrens commented on pavelib/servers.py:
```diff
@@ -48,40 +54,63 @@ def run_server(system, settings=None, port=None, skip_assets=False, contracts=Fa
 @needs('pavelib.prereqs.install_prereqs')
 @cmdopts([
     ("settings=", "s", "Django settings"),
+    ("asset-settings=", "a", "Settings file used for updating assets. Defaults to settings if not provided."),
```
"Defaults to settings if not provided." But only if settings aren't set... same note as above.

andy-armstrong replied:
The default behavior is that the asset_settings just follows the regular settings. It is only for the optimized case that the user wants to have a different settings file for asset generation than for running the server.

cahrens commented on pavelib/servers.py:
```diff
@@ -48,40 +54,63 @@ def run_server(system, settings=None, port=None, skip_assets=False, contracts=Fa
     port = getattr(options, 'port', None)
     fast = getattr(options, 'fast', False)
-    run_server('lms', settings=settings, port=port, skip_assets=fast)
+    collect_static = not fast and asset_settings != settings
```
Is the asset_settings != settings check correct? Do you mean asset_settings != DEFAULT_SETTINGS?

andy-armstrong replied:
This is what I intended (I did it in a few places). My thinking was that if the asset file was different than the regular settings file, that this meant that the user was generating static files that they were going to use. This isn't a perfect definition so maybe it would be better to make it an explicit parameter, but I don't imagine that anyone will use it.

jzoldak replied:
Might be worth a comment in the code.

andy-armstrong replied:
I've refactored the code a little so that the test is only done in two places, and I've gone with @cahrens's suggested check as I think it makes more sense. I've put a detailed comment in each of the two places (I couldn't find a way to share the code between the two, unfortunately).

cahrens commented on pavelib/servers.py:
```diff
@@ -48,40 +54,63 @@ def run_server(system, settings=None, port=None, skip_assets=False, contracts=Fa
+    run_server(
+        'studio',
+        settings=settings,
+        asset_settings=asset_settings if not fast else None,
```
As mentioned above, collect_static be eliminated?

cahrens commented on pavelib/servers.py:
```diff
@@ -96,7 +132,19 @@ def devstack(args):
+    if args.optimized:
+        settings = OPTIMIZED_SETTINGS
+        asset_settings = OPTIMIZED_ASSETS_SETTINGS
+    collect_static = not args.fast and asset_settings != settings
```
Same question here.

cahrens commented on pavelib/servers.py:
```diff
@@ -116,33 +164,66 @@ def celery(options):
+        args = [
+            'lms', 'studio',
+            '--settings={}'.format(asset_settings),
+            '--skip-collect'
```
Should skip-collect always be set?

andy-armstrong replied:
Yes. This one is a bit subtle, so it deserves a comment. I factored out the calls to collect assets because they are invoked with possibly different settings files. I first call update_assets while skipping collect so that all the other stuff is done (SASS, Coffee etc) once.

cahrens commented on requirements/edx/base.txt:
```diff
@@ -48,9 +48,6 @@ lazy==1.1
 lxml==3.3.6
 mako==0.9.1
 Markdown==2.2.1
---allow-external meliae
```
Unused?

andy-armstrong replied:
This was an attempt to fix the broken GitHub dependencies locally, but I didn't mean to check it in. I'll remove it.

cahrens commented:
I'm still not understanding the relationship between collect_static and whether or not asset_settings is set. Otherwise, it's looking good. Thanks for all the great tests!

cahrens commented:
:+1:

jzoldak commented on pavelib/servers.py:
```diff
@@ -116,33 +156,68 @@ def celery(options):
+    ("asset_settings=", "a", "Django settings for updating assets for both LMS and Studio (defaults to settings)"),
     ("worker_settings=", "w", "Celery worker Django settings"),
     ("fast", "f", "Skip updating assets"),
+    ("optimized", "f", "Run with optimized assets"),
```
I think you want some other character here besides "f" which is used for "fast"

andy-armstrong replied:
Nice catch. That was bad cut-and-paste on my part. I'm going with ``o``.

jzoldak replied:
Actually, come to think of it, you probably don't even need a shortcut. It doesn't make sense to run just the install_prereqs task directly with a different asset settings file. Your call.

jzoldak commented on pavelib/utils/process.py:
```diff
@@ -41,6 +43,11 @@ def run_multi_processes(cmd_list, out_log=None, err_log=None):
         err_log_file = open(err_log, 'w')
         kwargs['stderr'] = err_log_file
 
+    if tasks.environment.dry_run:
```
What does this do? Add an explanation in a code comment.

jzoldak commented:
`paver devstack --help` still doesn't give help. :(
I think it might be due to the @consume_args decorator - it's also consuming the --help?

andy-armstrong commented:
@jzoldak As you thought, it does seem to be an issue with using ``@consume_args``. There are three tasks that do this so that they can support a required first parameter ('lms' or 'studio') and it seems that ``--help`` doesn't work for any of them.

I did a quick Google search and haven't found a simple way to solve this. I suppose we will have to manually accept help and try to parse it. It is a little strange that you have to still specify 'lms' or 'studio' before you can specify ``--help``, so maybe we should special case that.

Given that this isn't a new problem, can I just report this as a bug and let someone else tackle it? We need to get this story landed, and the paver pieces were just a nice to have that have grown into a lot more work than I'd expected.

andy-armstrong commented:
@jzoldak @cahrens I've addressed the next round of feedback, so please re-review.

jzoldak commented:
@andy-armstrong Re: --help. Ugh. Yeah go ahead and put in a TestEng ticket. 
I think the root cause of our difficulty leads back to the initial implementation which requires lms or studio, as an arg not kwarg.

andy-armstrong commented:
@jzoldak I've created the ticket: https://openedx.atlassian.net/browse/TE-963

andy-armstrong commented:
@jzoldak @cahrens I've addressed your feedback so let me know if I'm good to merge.

jzoldak commented:
@andy-armstrong I cannot successfully run `paver devstack lms --optimized` on a fresh devstack. When it gets to `pavelib.assets.update_assets`, the manage.py command is `python manage.py lms --settings=test_static_optimized preprocess_assets`. Which fails because it's trying to validate the back end. You might need to change the DB settings that flow through the settings.py includes?
```
Traceback (most recent call last):
  File "manage.py", line 116, in <module>
    execute_from_command_line([sys.argv[0]] + django_args)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 443, in execute_from_command_line
    utility.execute()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 382, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/base.py", line 196, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/base.py", line 231, in execute
    self.validate()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/base.py", line 266, in validate
    num_errors = get_validation_errors(s, app)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/validation.py", line 103, in get_validation_errors
    connection.validation.validate_field(e, opts, f)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/mysql/validation.py", line 14, in validate_field
    db_version = self.connection.get_server_version()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 415, in get_server_version
    self.cursor().close()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/__init__.py", line 317, in cursor
    cursor = self.make_debug_cursor(self._cursor())
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 387, in _cursor
    self.connection = Database.connect(**kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/MySQLdb/__init__.py", line 81, in Connect
    return Connection(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/MySQLdb/connections.py", line 187, in __init__
    super(Connection, self).__init__(*args, **kwargs2)
_mysql_exceptions.OperationalError: (1049, "Unknown database 'edxtest'")
```

andy-armstrong commented:
@jzoldak Optimized isn't supported yet for LMS, as that is coming with this PR: https://github.com/edx/edx-platform/pull/8657. Can we just leave it like this for a few days until the other PR lands, or should I do some temporary interim solution?

jzoldak commented:
@andy-armstrong I get the same mysql exception when running `paver devstack studio --optimized`

andy-armstrong commented:
@jzoldak Thanks for catching this. When I originally wrote the ``test_static_optimized.py`` file, I based it upon the Bok Choy configuration. It turns out that it only works if Bok Choy has been run at least once before. The Bok Choy configuration steps are not necessary, as this settings file is only used to call collectstatic, so I've just removed all of those aspects. It continues to work on my machine, but please try it out on your clean devstack and let me know how it works.

andy-armstrong commented:
@jzoldak Uggh, that completely broke all the Bok Choy tests. I'm not sure why yet, but I'm in meetings until this afternoon so it will have to wait.

andy-armstrong commented:
@jzoldak The in-memory database solution worked, thanks! I've squashed the commits so that we can merge this in the morning, once you're comfortable with it.

FYI @cahrens, if you want to re-review.

cahrens commented:
I'm not sure there is a point in me giving it another thumbs-up... I haven't been very helpful thus far. But I did look it over again. :+1:

jzoldak commented:
:+1: :rocket: 
Quick, merge it @andy-armstrong :)

cahrens commented on lms/envs/devstack_optimized.py:
```diff
@@ -0,0 +1,40 @@
+
+  paver devstack lms --optimized
+
+You can also generate the assets explicitly and then run Studio:
```
@andy-armstrong should not reference Studio.
